### PR TITLE
Object Tagging support

### PR DESCRIPTION
**Pull request by jagregory**
Still a work-in-progress.

Adds Object Tagging support. PutObjectTagging saves a `key.tagging.xml` subresource in the data directory containing the tags, and GetObjectTagging reads that file.

### CODE REVIEW

File: lib/stores/filesystem.js [hunk 1]
```diff
@@ -305,6 +326,35 @@ class FilesystemStore {
     return { size, md5 };
   }
 
+  async putObjectTagging(bucket, key, content) {
+    const taggingPath = this.getResourcePath(bucket, key, "tagging.json");
+    const parsedTagging = await parseXmlBody(content);
```
Comment (jagregory) [inline lib/stores/filesystem.js]: @kherock Not sure what's the preferred approach here. The SDK sends tags as XML in the request body, I parse this and then save it to JSON, and then on the way out I do the reverse. I'm wondering if I should just take the payload, save it, and then send it back out again for Get requests without touching it at all?

Comment (kherock) [reply lib/stores/filesystem.js]: That's the approach I would have tried first - you should try seeing if you can reuse `pubSubresource` and `getSubresource` for this. However, those methods are defined to use a config model (such as ones we have for website and CORS configurations) that would parse and validate tags. Let me know if you need more guidance on that.

**Comment by jagregory** [comment]
@kherock I redid this using Subresources and it's a lot simpler, thanks! I reckon this is good for a review when you have time.

File: lib/controllers/object.js [hunk 2]
```diff
@@ -354,6 +355,57 @@ exports.putObjectCopy = async function putObjectCopy(ctx) {
...
+    ctx.logger.error(
+      'Error tagging object "%s" in bucket "%s"',
+      key,
+      bucket,
+      err
+    );
+    throw err;
+  }
+};
+
```
Comment (kherock) [inline lib/controllers/object.js]: Could you insert these definitions in the order as they appear in this list here?
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectOps.html

The structure under `controllers/` is meant to mirror the official REST API docs.

File: lib/utils.js [hunk 3]
```diff
@@ -120,3 +120,19 @@ exports.parseDate = function(dateString) {
 exports.toISO8601String = function(date) {
   return new Date(date).toISOString().replace(/[-:]|\.\d+/g, "");
 };
+
+/**
+ * Reads a request body stream to a string.
+ *
+ * @param ctx
+ * @returns {Promise<string>}
```
Comment (kherock) [inline lib/utils.js]: This is actually `Promise<void>`, we're not returning anything. You can also give `ctx` a type of `Koa.Context` if you'd like (there's no TS compatibility yet so it doesn't matter too much to me).

File: lib/models/config.js [hunk 4]
```diff
@@ -341,3 +343,26 @@ class S3WebsiteConfiguration extends S3ConfigBase {
...
+    }
+
+    return new TaggingConfiguration(xml);
+  }
+
+  constructor(config) {
+    super("tagging", config);
+  }
+}
+TaggingConfiguration.empty = new TaggingConfiguration(
```
Comment (kherock) [inline lib/models/config.js]: How about naming this `TaggingConfiguration.EMPTY`?

**Comment by jagregory** [comment]
Changes pushed :)

I've also added a couple more tests for error scenarios too.

**Comment by jagregory** [comment]
Hey @kherock, anything else needed to merge this? Cheers

**Comment by kherock** [comment]
This is released in v3.4.0! Sorry for the wait.

**Comment by jagregory** [comment]
All good, thanks! 🙂

**Pull request merged**